### ISSUE TITLE
Data.Semigroup.Foldable: Can't be safely imported! (5.2.2)

### ISSUE THREAD

**Issue by phadej** [ekmett/semigroupoids#69]
```
src/Algebra/Lattice.hs:74:1: error:
    Data.Semigroup.Foldable: Can't be safely imported!
    The module itself isn't safe.
```

I suspect it's due

```diff
 import Data.Orphans ()
+import Language.Haskell.TH (Q)
 import Prelude hiding (id, (.))
```

---

This is just a heads-up, the failures may show up somewhere or may not.

**Comment by RyanGlScott** [comment]
What is `src/Algebra/Lattice.hs`?

I did turn off `Trustworthy` in some modules in `semigroupoids-5.2.2`, but was hoping that this wouldn't actually break anything (alas, it seems I wrong wrong on that front).

**Comment by phadej** [comment]
`src/Algebra/Lattice.hs` is a module in `lattices`

**Comment by phadej** [comment]
Anyway, maybe we should add a note to changelog, otherwise there's nothing to do.

I think this issue can be closed: if people search for the issues, they will hopefully find this, and adopt.

**Comment by RyanGlScott** [comment]
Actually, a correction to my comment in https://github.com/ekmett/semigroupoids/issues/69#issuecomment-361295111: I wasn't the one who changed the Safe Haskell behavior, it was an unintended consequence of https://github.com/ekmett/semigroupoids/commit/6d981d806476db1e1a1ad80dee970fb1bc8b9835 (as noted in https://github.com/ekmett/pointed/issues/17#issuecomment-359432778). I don't think I'm in violation of the PVP here, since it doesn't say anything about Safe Haskell, but I should probably add a note to the `CHANGELOG` retroactively about this.

**Comment by phadej** [comment]
Yes, PVP doesn't say anything about *Safe Haskell*, so it's fine. Relying on implicit `Safe` is crying for troubles, as we have repeatedly experienced.

**Comment by phadej** [comment]
To be clear: there is no intention to mark `Data.Semigroup.Foldable` as `Trustworthy` for older GHC? So it will stay *Unsafe*?

**Comment by RyanGlScott** [comment]
Someone other modules in `semigroupoids` attempt to mark things as `Trusthworthy`—for example, [here](https://github.com/ekmett/semigroupoids/blob/6d981d806476db1e1a1ad80dee970fb1bc8b9835/src/Data/Functor/Bind/Class.hs#L7-L9). But as you can see, getting it to work is extremely fiddly, as you must figure out which range of dependency versions are in turn `Trustworthy` or `Safe`. Plus, if you go too far in the other direction and just mark everything as `Trustworthy` unconditionally, GHC might warn that you have modules marked as `Trustworthy` but inferred as `Safe`! So unless you get the CPP _just_ right, GHC will bark at you in some way.

To reiterate my feelings on the matter from https://github.com/ekmett/pointed/issues/17#issuecomment-359485775: I do not feel motivated whatsoever to try as figure out the right incantation for making everything in `semigroupoids` safely importable. If you feel up to the task, I'd accept a pull request, but under the condition that I'll shift the blame on you if anything related to Safe Haskell breaks in the future ;)

**Comment by RyanGlScott** [comment]
I've updated the `CHANGELOG` in https://github.com/ekmett/semigroupoids/commit/303774ecc08fb3c8ac12519a38caaa6c521d0aa1, so I'll close this issue.

**Comment by phadej** [comment]
@RyanGlScott 

![screenshot from 2018-01-29 19-46-30](https://user-images.githubusercontent.com/51087/35525464-3f046356-052d-11e8-88c9-f27f77cdbb5f.png)

I think keeping all up to `semigroupoids` and `profunctors` Safe/Trustworthy is doable (that's dep graph with GHC-7.8.4), there's finite amount modules.

To my understanding if everything is tagged as either `Trustworthy` or `Safe`, than the worst thing which may happen is that we get "this module is marked as Trustworthy though inferred as Safe" if some dependency becomes `Safe` itself.

I can undergo the mission of making `semigroupoids` safe, as I have a tooling to sweep through dependency ranges with high confidence.

FWIW latest `comonad` fails to compile with GHC-7.2.2 (because `semigroups-0.18.3` cannot be safely imported)

---

TL;DR to make `semigroupoids` Safe, we have to work from the bottom of dependency tree. Non-systematic approach is doomed to fail

**Comment by RyanGlScott** [comment]
Alright, you're in charge :)

**Comment by phadej** [comment]
@RyanGlScott, ok, but don't hold your breath

**Comment by RyanGlScott** [comment]
As the discussion on https://github.com/haskell/cabal/issues/5010 revealed, there's another Safe Haskell issue that's commonplace in many Kmett libraries: namely, that `Safe` is used at all on GHC 7.2, since 7.2's treatment of Safe Haskell is [completely broken](https://github.com/haskell/cabal/issues/5010#issuecomment-362130386). The only apparent fix is to downgrade from `Safe` to `Trustworthy` on 7.2.

**Comment by RyanGlScott** [comment]
After doing a quick search, here's a list of Kmett libraries I've found that enable `Safe` on GHC 7.2:

* [ ] [`bifunctors`](https://github.com/ekmett/bifunctors/blob/9131271ee13628e768f2a4a1d3c6b4006a22c2cc/src/Data/Bifunctor/Functor.hs#L5-L7)
* [ ] [`comonad`](https://github.com/ekmett/comonad/blob/cd57e2bf6b84d25bcd45c7e7cf76a9f07d71e381/src/Control/Comonad/Store/Class.hs#L6-L8)
* [ ] [`nats`](https://github.com/ekmett/nats/blob/ccd6796cfc0ece32fb682cf69590350b5672e2bd/src/Numeric/Natural.hs#L12-L18)
* [ ] [`transformers-compat`](https://github.com/ekmett/transformers-compat/blob/2faad6e6ad257e8ca8874fef0d636a08fc83d7dc/0.3/Control/Monad/Signatures.hs#L4-L6)

This might not be exhaustive, but it's at least a good place to start.

**Comment by RyanGlScott** [comment]
I've published new versions of (most of) the packages in https://github.com/ekmett/semigroupoids/issues/69#issuecomment-363444324 that do not enable `Safe` on GHC 7.2.

**Comment by phadej** [comment]
This causes lens failures:

![screenshot from 2018-02-12 00-52-48](https://user-images.githubusercontent.com/51087/36095790-53e65f84-0ffc-11e8-97f6-3e8923b4c5fc.png)

I can make mass update to make `semigroupoids <6` -> `semigroupoids <5.2.2`. 

`lens-4.13.2.1`  *Uploaded Fri Feb 26 18:26:37 UTC 2016*, so having upper bound of 5.2.2 shouldn't affect e.g. Stackage or other freeze files.

ping @hvr

**Comment by hvr** [comment]
@phadej sounds safe... erm good to me :-)

![safe](https://user-images.githubusercontent.com/285533/35805654-015ced16-0a7d-11e8-8b03-b81e5a43fe8a.jpeg)

**Comment by ekmett** [comment]
> Relying on implicit `Safe` is crying for troubles, as we have repeatedly experienced.

The problem is that relying on _explicit_ `Safe` is even worse as it turns an issue 99% of folks don't care about into an actual compile time error. =(

Up until now I've been sprinkling `Trustworthy` wherever I use a known-unsafe feature in a safe way, but its an incredibly fallible approach.

I'm starting to lean towards the approach I'm piloting in `profunctors`.

```
if impl(ghc>=8.0)
    ghc-options: -Wno-trustworthy-safe
```

Then I can feel free to spam `Trustworthy` more liberally, using it everywhere I ever see unsafety and know I'm not the culprit.

`-fpackage-trust` is the only thing that cares about `Trustworthy`, and that works on a package level of granularity, so if we have any `Trustworthy` in the package, then having even more of it, just means its harder to audit sanely for security.

I really want a `TrustworthyOrBetter` language pragma. The status quo is awful.

**Comment by RyanGlScott** [comment]
I believe #97 has improved the Safe Haskell support in `semigroupoids` to the point where we no longer need to keep this issue open. I'll close this, but please re-open if you feel that there is something else that needs to be addressed here.
